### PR TITLE
pruner export

### PR DESCRIPTION
**Pull request by tanglang96**

### CODE REVIEW

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 1]
```diff
@@ -114,6 +115,48 @@ def new_forward(*inputs):
...
+            if mask is not None:
+                mask_sum = mask.sum().item()
+                mask_num = mask.numel()
+                _logger.info('Layer: {}  Sparsity: {}'.format(name, 1 - mask_sum / mask_num))
+                print('Layer: {}  Sparsity: {}'.format(name, 1 - mask_sum / mask_num))
+                m.weight.data = m.weight.data.mul(mask)
+            else:
+                _logger.info('Layer: {}  Sparsity: {}'.format(name, 0))
+                print('Layer: {}  Sparsity: {}'.format(name, 0))
+        assert save_path is not None, 'Save Path must be specified'
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: suggest to separate mask and model to two files

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 2]
```diff
@@ -114,6 +115,48 @@ def new_forward(*inputs):
 
         layer.module.forward = new_forward
 
+    def export_model(self, model, save_path=None, onnx_path=None, input_shape=None):
+        """Pruned model can be exported with `export_model`.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: please follow our docstring format

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 3]
```diff
@@ -114,6 +115,48 @@ def new_forward(*inputs):
...
+        `onnx_path` can be optionally specified for saving model to `.onnx` format.
+        `input_shape` is needed for exporting model to `.onnx` format.
+        """
+        export_dict = {}
+        export_dict['state_dict'] = model.state_dict()
+        mask_dict = {}
+        for name, value in model.state_dict():
+            found = False
+            for layer_name, mask in self.mask_list:
+                if name == 'module.' + layer_name + '.weight' or layer_name + '.weight':
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: This if sentence seems incorrect.

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 1 above)
Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: @QuanluZhang actually we often save checkpoints by saving optimizer state_dict and model state_dict to one file while using pytorch, here we can get the dicts like this
```
prune_dict = torch.load('prune_dict.pth')
state_dict = prune_dict['state_dict']
mask_dict = prune_dict['mask_dict']
```

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 2 above)
Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: @QuanluZhang I'll fix this later

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 3 above)
Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: @chicm-ms I've fixed this.

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 4]
```diff
@@ -114,6 +115,50 @@ def new_forward(*inputs):
...
+            if mask is not None:
+                mask_sum = mask.sum().item()
+                mask_num = mask.numel()
+                _logger.info('Layer: {}  Sparsity: {}'.format(name, 1 - mask_sum / mask_num))
+                print('Layer: {}  Sparsity: {}'.format(name, 1 - mask_sum / mask_num))
+                m.weight.data = m.weight.data.mul(mask)
+            else:
+                _logger.info('Layer: {}  Sparsity: {}'.format(name, 0))
+                print('Layer: {}  Sparsity: {}'.format(name, 0))
+        assert save_path is not None, 'Save Path must be specified'
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: better move this input checking assertion to the begining of the function.

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 5]
```diff
@@ -114,6 +115,50 @@ def new_forward(*inputs):
...
+        onnx_path : str
+            (optional) path to save onnx model
+        input_shape : list or tuple
+            input shape to onnx model
+        """
+        export_dict = {}
+        export_dict['state_dict'] = model.state_dict()
+        mask_dict = {}
+        for layer_name in self.mask_list:
+            mask_dict[layer_name + '.weight'] = self.mask_list[layer_name]
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: It is a little be wield to do this conversion here, normally we do not call a mask as weight.   And if we need to do the conversion, better to save it as 'xxx.weight' at compression time.

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 4 above)
Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: @chicm-ms I've fixed this.

File: src/sdk/pynni/nni/compression/torch/compressor.py (hunk 5 above)
Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: @chicm-ms I've fixed, using a dict to store masks, users can get mask like this
 ```mask = mask_dict[layer_name]['weight']```

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 6]
```diff
@@ -114,6 +115,51 @@ def new_forward(*inputs):
...
+        input_shape : list or tuple
+            input shape to onnx model
+        """
+        assert save_path is not None, 'Save Path must be specified'
+        export_dict = {}
+        export_dict['state_dict'] = model.state_dict()
+        mask_dict = {}
+        for layer_name in self.mask_list:
+            mask_dict[layer_name] = {}
+            mask_dict[layer_name]['weight'] = self.mask_list[layer_name]
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: mask_list does not exist in this class

File: src/sdk/pynni/nni/compression/torch/builtin_pruners.py [hunk 7]
```diff
@@ -17,7 +17,7 @@ def __init__(self, model, config_list):
             - sparsity
         """
         super().__init__(model, config_list)
-        self.mask_list = {}
+        self.mask_dict = {}
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/builtin_pruners.py]: `self.mask_dict` already exists in `Pruner` class

Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/builtin_pruners.py]: @QuanluZhang fixed.

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 8]
```diff
@@ -177,6 +188,48 @@ def new_forward(*inputs):
...
+        for name, m in self.bound_model.named_modules():
+            mask = self.mask_dict.get(name)
+            if mask is not None:
+                mask_sum = mask.sum().item()
+                mask_num = mask.numel()
+                _logger.info('Layer: %s  Sparsity: %.2f', name, 1 - mask_sum / mask_num)
+                print('Layer: %s  Sparsity: %.2f' % (name, 1 - mask_sum / mask_num))
+                m.weight.data = m.weight.data.mul(mask)
+            else:
+                _logger.info('Layer: %s  Sparsity: %.2f', name, 0)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: it would be better to show this layer is not compressed, for example: 'Not compressed Layer: .....'

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 9]
```diff
@@ -177,6 +188,48 @@ def new_forward(*inputs):
...
+                print('Layer: %s  Sparsity: %.2f' % (name, 0))
+        torch.save(self.bound_model.state_dict(), model_path)
+        _logger.info('Model state_dict saved to %s', (model_path))
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: remove bracket

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 10]
```diff
@@ -128,11 +128,22 @@ def _expand_config_op_types(self, config):
                 expanded_op_types.append(op_type)
         return expanded_op_types
 
+
 class Pruner(Compressor):
-    """
-    Abstract base PyTorch pruner
+    """Prune to an exact pruning level specification
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: newline after quotes

Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/compressor.py]: @QuanluZhang all fixed.

File: src/sdk/pynni/nni/compression/torch/compressor.py [hunk 11]
```diff
@@ -177,6 +189,48 @@ def new_forward(*inputs):
...
+        model_path : str
+            path to save pruned model state_dict
+        mask_path : str
+            (optional) path to save mask dict
+        onnx_path : str
+            (optional) path to save onnx model
+        input_shape : list or tuple
+            input shape to onnx model
+        """
+        assert model_path is not None, 'model_path must be specified'
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/compressor.py]: If model_path can not be None,  we should not provide None as default value.
